Scanning a monorepo in parts (#1299)
semgrep/semgrep-docs: docs/kb/semgrep-ci/scan-monorepo-in-parts.md

---
description: How to scan a monorepo in parts for better CI performance and clearer findings organization
tags:
  - monorepo
  - optimization
  - performance
  - ci
---

# Scanning a monorepo in parts

Monorepos, by definition, contain a large amount of code. The code is also often divided into to different components or modules, despite all existing in the same repository.

> As such, it can be helpful to scan a monorepo in parts for multiple reasons:

Author: I feel like this suggestion is just repeating what I mentioned above it:
> Monorepos very often contain a large amount of code. The code is usually divided into to different components or modules, even though it's all stored in the same repository.

Which is why I think "As such" makes sense, it is continuing that same thought process

> 1. Improving scan performance in CI and reducing CI run times
> 2. Logically splitting up the monorepo into multiple components to help with managing findings

Reviewer: ```suggestion
* Improve scan performance in CI and reduce CI run times
* Logically split the monorepo to simplify managing findings
```

Author: Waiting on the other convo above to decide how to change this

## How to configure Semgrep in CI to split up a monorepo

> Typically when scanning any repo with Semgrep in CI, the base command is `semgrep ci`. Of course, you need to have an auth token set and perhaps other environment variables configured depending on what SCM and CI provider you are using, but this article will assume you are already familiar with those settings.

Reviewer: ```suggestion
When scanning a repo with Semgrep in CI, the base command is `semgrep ci`. To understand this default setup for your source code manager (SCM) and CI provider, see [Getting started with Semgrep in continuous integration (CI)](/docs/semgrep-ci/overview/).
```

> To split up your monorepo, you need to first decide *how* you want to logically split up the code. For example, if your monorepo has four main modules and their paths are:

Reviewer: ```suggestion
To split up your monorepo, first decide *how* you want to logically split up the code. For example, if the monorepo has four main modules and their paths are:
```

    /src/moduleA
    /src/moduleB
    /src/moduleC
    /src/moduleD

> Then it would logically make sense to split this into four separate scans in CI, one for each module.

Reviewer: ```suggestion
Then splitting its scans into four separate scans, one for each module, would provide a logical separation for findings.
```

> To actually achieve this, you would use the `--include` flag ([see CLI reference](/docs/cli-reference/)) to only scan files in each module's code path:

Reviewer: ```suggestion
After choosing a logical split, use the `--include` flag ([see CLI reference](/docs/cli-reference/)) with the relevant path to only scan files in that module's code path:
```

    semgrep ci --include=/src/moduleA/*

> Now, Semgrep is only scanning files under that path and the entire CI run will take less time, since less code is being scanned. Of course, you will now have to run multiple CI runs for the remaining three modules to get full monorepo coverage, but you have more flexibility on how and when you perform those runs.

Reviewer: I think I'd go ahead and add all four commands under the first one here, something like this:

> For the other modules, the commands look similar:
> 
> &lt;block of commands>
> 
> Each CI run will take less time, and you will have the flexibility to trigger each one on appropriate events or frequencies.
> 

(this isn't ideal wording, here I'm just giving an idea of a different approach)

Author: I'm not sure, I think that might make the article unnecessarily convoluted. It's clear they need to run this for the other modules, but when and how they do so is up to them.

> Now that you've successfully configured your monorepo to be scanned in parts, you also have to configure the findings from each part or module to show up as their own project in Semgrep Cloud Platform (SCP).

Reviewer: I'd still suggest integrating this further, since we talk about separating findings as part of the benefit at the very top. Maybe something like changing line 32 above to:

> After choosing a logical split, use the `--include` flag ([see CLI reference](/docs/cli-reference/)) with the relevant path to only scan files in that module's code path, and add the the `SEMGREP_REPO_NAME` environment variable ([see CI environment variables reference](/docs/semgrep-ci/ci-environment-variables/#semgrep_repo_name)) to send the findings to a different project in Semgrep.

Author: I do get the idea behind this, but I also think it muddies up the 2 concepts, 1) being dividing up the paths with --include and 2) being sending the findings to separate projects. I'm afraid if I write something similar to your suggestion, it will distract from just getting part 1) right. It makes sense in my mind to have them as 2 separate parts but emphasize that both are needed for this whole thing to work.

For example, if your monorepo is located at `https://github.com/sina/monorepo` the `SEMGREP_REPO_NAME` would typically be set to `sina/monorepo`. So to split it up into the four modules we referenced above, for each CI run of a module we need to manually set it like so before running Semgrep:

    export SEMGREP_REPO_NAME="sina/monorepo/moduleA"

And then running Semgrep as demonstrated above:

    semgrep ci --include=/src/moduleA/*

Now, the findings from this CI run will show up in their own project in SCP named `sina/monorepo/moduleA`. This is not only necessary when splitting up a monorepo but also helpful in terms of organizing findings into separate projects so that devs and security engineers can have a clearer understanding of which findings pertain to the module that they are actually responsible for.

:::info
Due to how we determine hyperlinks for findings in SCP, changing the `SEMGREP_REPO_NAME` to anything other than the actual `<org>/<repo name>` format may adversely affect and break hyperlinks. This is a tradeoff of splitting up one single repo into multiple projects.

> :::

Reviewer: I'd move this up as well, likely keeping it as an info box wherever you first introduce the SEMGREP_REPO_NAME usage. I'd write it a bit differently though:

> Changing the `SEMGREP_REPO_NAME` value in a scan so that it does not match the repo's `<org>/<repo name>` value on the SCM may cause issues with code hyperlinks in Semgrep Cloud Platform. This is a necessary tradeoff when splitting up a repo into multiple projects.

---
description: How to scan a monorepo in parts for better CI performance and clearer findings organization
tags:
  - monorepo
  - performance
  - Semgrep in CI
---

# Scanning a monorepo in parts

With default CI configurations, monorepos will be scanned as a single project in Semgrep. However, monorepos very often contain a large amount of code and the code is usually divided into to different components or modules.

As such, it can be helpful to scan a monorepo in parts for multiple reasons:

* To improve scan performance in CI and reduce CI run times
* To logically split the monorepo to simplify managing findings

## How to configure Semgrep in CI to split up a monorepo

When scanning a repo with Semgrep in CI, the base command is `semgrep ci`. To understand this default setup for your source code manager (SCM) and CI provider, see [Getting started with Semgrep in continuous integration (CI)](/docs/semgrep-ci/overview/).

> To split up your monorepo, you will need to make 2 changes. First, using the `--include` flag to determine *how* you want to logically split up the code. Second, updating the `SEMGREP_REPO_NAME` environment variable to assign findings to separate projects in Semgrep Cloud Platform (SCP). 

Reviewer: ```suggestion
To split up your monorepo, you need to make two changes. First, use the `--include` flag to determine *how* you want to logically split up the code. Second, update the `SEMGREP_REPO_NAME` environment variable to assign findings to separate projects in Semgrep Cloud Platform (SCP). 
```

For example, if the monorepo has four main modules and their paths are:

    /src/moduleA
    /src/moduleB
    /src/moduleC
    /src/moduleD

Then splitting its scans into four separate scans, one for each module, would provide a logical separation for findings.

After choosing a logical split, use the `--include` flag ([see CLI reference](/docs/cli-reference/)) with the relevant path to only scan files in that module's code path:

    semgrep ci --include=/src/moduleA/*

> Now, Semgrep is only scanning files under that path and the entire CI run will take less time, since less code is being scanned.

Reviewer: ```suggestion
Now, Semgrep is only scanning files under that path, and the CI run will take less time, since less code is being scanned.
```

For the other modules, the commands look similar. For module B:

    semgrep ci --include=/src/moduleB/*

You will then have the flexibility to trigger each one on appropriate events or frequencies.

> Now that you've successfully configured your monorepo to be scanned in parts, you also have to configure the findings from each part or module to show up as their own project in SCP.

Reviewer: ```suggestion
Now that you've successfully configured your monorepo to be scanned in parts, move on to configuring the findings from each part or module to show up as their own project in SCP.
```

To ensure findings from the module are assigned to their own project in SCP, you will need to explicitly set the `SEMGREP_REPO_NAME` environment variable ([see CI environment variables reference](/docs/semgrep-ci/ci-environment-variables/#semgrep_repo_name)).

:::info

> Changing the `SEMGREP_REPO_NAME` value in a scan so that it does not match the repo's actual `<org>/<repo name>` structure on the SCM may cause issues with code hyperlinks in Semgrep Cloud Platform. This is a necessary tradeoff when splitting up a repo into multiple projects.

Reviewer: ```suggestion
Changing the `SEMGREP_REPO_NAME` value in a scan so that it does not match the repo's `<org>/<repo name>` structure on the SCM may cause issues with code hyperlinks in Semgrep Cloud Platform. This is a necessary tradeoff when splitting up a repo into multiple projects.
```

:::

> For example, if your monorepo is located at `https://github.com/semgrep/monorepo` the `SEMGREP_REPO_NAME` would typically be set to `semgrep/monorepo`. So to split it up into the four modules we referenced above, for each CI run of a module we need to manually set it like so before running Semgrep:

Reviewer: ```suggestion
For example, if your monorepo is located at `https://github.com/semgrep/monorepo` the `SEMGREP_REPO_NAME` would typically be set to `semgrep/monorepo`. To split the single project into four projects corresponding to the logical modules, set `SEMGREP_REPO_NAME` to match the module name before running Semgrep:
```

    export SEMGREP_REPO_NAME="semgrep/monorepo/moduleA"

And then running Semgrep as demonstrated above:

    semgrep ci --include=/src/moduleA/*

> Now, the findings from this CI run will show up in their own project in SCP named `semgrep/monorepo/moduleA`. This is not only necessary when splitting up a monorepo but also helpful in terms of organizing findings into separate projects so that devs and security engineers can have a clearer understanding of which findings pertain to the module that they are actually responsible for.

Reviewer: ```suggestion
Now, the findings from this CI run will show up in their own project in SCP named `semgrep/monorepo/moduleA`. This is not only necessary to ensure findings have a consistent status, but also helpful so that developers and security engineers can have a clearer understanding of which findings pertain to the module that they are responsible for.
```
